clients/drcachesim/tracer/func_trace.cpp

```diff
@@ -0,0 +1,256 @@
+split_by(std::string s, std::string sep)
+{
+    size_t pos;
+    std::vector<std::string> vec;
+    do {
+        pos = s.find(sep);
+        vec.push_back(s.substr(0, pos));
+        s.erase(0, pos + sep.length());
+    }
+    while (pos != std::string::npos);
+    return vec;
+}
+
+static void
+get_funcs_str_and_sep(std::string &funcs_str, std::string &sep)
+{
+    if (op_record_heap.get_value())
+        funcs_str = op_record_heap_value.get_value();
+    sep = op_record_function.get_value_separator();
+    DR_ASSERT(sep == op_record_heap_value.get_value_separator());
+    std::string op_value = op_record_function.get_value();
+    if (!funcs_str.empty() && !op_value.empty())
+        funcs_str += sep;
+    funcs_str += op_value;
+}
+
+DR_EXPORT
+bool
+func_trace_init(func_trace_append_entry_t append_entry_)
+{
+    if (dr_atomic_add32_return_sum(&func_trace_init_count, 1) > 1)
+        return true;
+
+    std::string funcs_str, sep;
+    get_funcs_str_and_sep(funcs_str, sep);
+    // If there is no function specified to trace,
+    // then the whole func_trace module doesn't have to do anything.
+    if (funcs_str.empty())
+        return true;
+
+    auto op_values = split_by(funcs_str, sep);
+    std::set<int> existing_ids;
+    if (!drvector_init(&funcs, op_values.size(), false, free_func_entry)) {
+        DR_ASSERT(false);
+        goto failed;
+    }
+    append_entry = append_entry_;
+
+    for (auto &single_op_value : op_values) {
+        auto items = split_by(single_op_value, PATTERN_SEPARATOR);
+        if (items.size() != 3) {
+            NOTIFY(0, "Warning: -record_function or -record_heap_value was not"
+                      " passed a triplet, funcs_str=%s\n", funcs_str.c_str());
+            continue;
+        }
+        std::string name = items[0];
+        int id = atoi(items[1].c_str());
+        int arg_num = atoi(items[2].c_str());
+        if (existing_ids.find(id) != existing_ids.end()) {
+            NOTIFY(0, "Warning: duplicated function id in -record_function or"
+                      " -record_heap_value, funcs_str=%s\n", funcs_str.c_str());
+            continue;
+        }
+        dr_log(NULL, DR_LOG_ALL, 1, "Trace func name=%s, id=%d, arg_num=%d\n",
+               name.c_str(), id, arg_num);
+        existing_ids.insert(id);
+        drvector_append(&funcs, create_func_metadata(name, id, arg_num));
+    }
+
+    if (!(drsym_init(0) == DRSYM_SUCCESS)) {
+        DR_ASSERT(false);
+        goto failed;
+    }
+    if (!drwrap_init()) {
+        DR_ASSERT(false);
+        goto failed;
+    }
+    if (!drmgr_register_module_load_event(instru_funcs_module_load)) {
+        DR_ASSERT(false);
+        goto failed;
+    }
+
+    return true;
+failed:
+    func_trace_exit();
+    return false;
+}
+
+DR_EXPORT
+void
+func_trace_exit()
+{
+    if (dr_atomic_add32_return_sum(&func_trace_init_count, -1) != 0)
+        return;
+
+    std::string funcs_str, sep;
+    get_funcs_str_and_sep(funcs_str, sep);
```

Comment:
I think it's best to cache the function string during init and avoid this parsing here: there are scenarios where using STL code at exit can cause problems (not in google3 usage, but in the dr_app_take_over() scenario where the app exits before DR).

Reply:
Done. Ready for next round of review!